File: cpp/src/Test8.cpp
#include "TestHelpers.h"
#include "Tree.h"
#include "Problem8.h"
#include <cstdio>
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <string>



class Test {
private:
static int getResult(std::string& f) {
     std::string resultsFileName = TestHelpers::getResultsFile(f);
     std::ifstream ifs(resultsFileName);
     int res;

     std::string data;
     if (ifs) {
         ifs >> data;
         if (std::isdigit(data[0])) {
             res = std::stoi(data);
         } else {
	   res = -1;
         }
     } else {
         std::cout << "Results file not found: " << resultsFileName << std::endl;;
         exit(1);
     }
     return res;
 }

static std::vector<Tree*> getTree(std::string& f) {
  std::string dataFileName = TestHelpers::getDataFile(f);

    int num_nodes;
    std::ifstream ifs(dataFileName);
    if (ifs) {
        ifs >> num_nodes;
    } else {
        std::cout << dataFileName << " failed to be read" << std::endl;;
        return std::vector<Tree*>();
    }
    
    std::vector<Tree*> nodes(num_nodes+1, nullptr);
    for (int i = 0; i < num_nodes; ++i) {
        int num, salary;
        ifs >> num >> salary;
        nodes[num] = new Tree(num, salary);
    }
    
    for (int i = 0; i < num_nodes - 1; ++i) {
        int source, sink;
        ifs >> source >> sink;
        if (source >= num_nodes+1 || sink >= num_nodes+1) {
        } else {
            nodes[source]->addDirectReport(nodes[sink]);
        }
        
    }

    return nodes;
}

public:
static bool runTest(std::string f) {
    std::string dataFileName = TestHelpers::getDataFile(f);
    Tree* root = nullptr;
    std::vector<Tree*> nodes = getTree(f);
    int expected = getResult(f);
    int actual = -1;

    if ((nodes.size() == 0) || (expected < 0)) {
      return false;
    }

    actual = Problem8::solve(nodes[1]);

    if (actual == expected) {
        return true;
    } else {
        std::cout << "Failed actual result and expected result do not match" << "\n";
        std::cout << "    actual is " << actual << "\n";
        std::cout << "    expected is " << expected << "\n";
	return false;
    }
}

};

int main (int argc, char* argv[]) {
    std::vector<std::string> files = TestHelpers::getFiles(argc, argv);
    if (files.size() == 0) {
        std::cout << "no test file found"; 
        exit(1);
    } 

    int count = 0;
    for (size_t i = 0; i < files.size(); ++i) {
        std::string f = files[i];
	std::cout << "Test " << f << ": ";	 	
        if (Test::runTest(f)) {
	  std::cout << "PASSED\n";	  
	  count++;
	} else {
	  std::cout << "FAILED\n";	  
	}
    }

    if (count == files.size()) {
      return 42;
    } else {
      return 0;
    }


}

File: java/src/Test8.java
/* Solution to paths problem.
 * Version w/ integer assignment ids
 */

import java.util.Scanner;
import java.util.ArrayList;
import java.util.List;
import java.io.*;

public class Test8 extends Test {
    /**
     * runTest: run the test in the specified file.  Returns true, if the test passes
     *   and false otherwise.
     **/
    public boolean runTest(File f, boolean verbose) {
	Tree root = null;
	List<Tree> nodes = loadData(f, verbose);
	int expected = getResults(f, verbose);
	int actual = -1;

	if ((nodes == null) || (expected < 0)) {
	    return false;
	}

	try {
	    actual = Problem8.solve(nodes.get(1));
	} catch (Exception e) {
	    if (verbose) {
		System.out.println("Expected exception: " + e);
	    }
	    return false;
	}

	if (actual == expected) {
	    return true;
	} else {
	    if (verbose) {
		System.out.println("Failed actual result and expected result do not match");
		System.out.println(String.format("    actual is %d", actual));
		System.out.println(String.format("    expected is %d", expected));
	    }
	    return false;
	}
    }

    /**
     * loadData: load the company data into a Tree.
     **/
    private static List<Tree> loadData(File f, boolean verbose) {
	ArrayList<Tree> nodes = null;
	String[] line;
	// Read the data
	try {
	    Scanner scanner = new Scanner(f);
	    String[] line0 = scanner.nextLine().split(" ");
	    int num_nodes = Integer.parseInt(line0[0]);

	    nodes = new ArrayList<Tree>();
	    for (int i = 0; i < num_nodes + 1; i++) {
		nodes.add(null);
	    }

	    for (int i = 0; i < num_nodes; i++) {
		line = scanner.nextLine().split(" ");
		int num = Integer.parseInt(line[0]);
		int salary = Integer.parseInt(line[1]);
		nodes.set(num, new Tree(num, salary));
	    }

	    for (int i = 0; i < num_nodes-1; i++) {
		line = scanner.nextLine().split(" ");
		int source = Integer.parseInt(line[0]);
		int sink = Integer.parseInt(line[1]);
		nodes.get(source).addDirectReport(nodes.get(sink));
	    }
	} catch (FileNotFoundException e) {
	    if (verbose) {
		System.out.println("Results file not found:" + f.getPath());
	    }
	    return null;
	} catch (NumberFormatException e) {
	    if (verbose) {
		System.out.println("Format error: line 0");
	    }
	    return null;
	} catch (Exception e) {
	    if (verbose) {
		System.out.println("Unexpected exception: " + e);
	    }
	    return null;
	}
	return nodes;
    }

    /**
     * getResults: get the expected results from the results file.
     **/
    private static Integer getResults(File inputFile, boolean verbose) {
	Integer result = -1;
	Scanner scanner = TestHelpers.getResultsFileScanner(inputFile);

	if (scanner == null) {
	    return result;
	}

	try {
	    String data = scanner.nextLine();
	    result = Integer.parseInt(data);
	} catch (NumberFormatException e) {
	    if (verbose) {
		System.out.println("Results file has the wrong format");
	    }
	    return -1;
	} catch (Exception e) {
	    if (verbose) {
		System.out.println("Expected exceptions: " + e);
	    }
	    return -1;
	}
	return result;
    }
    
    public static void main(String[] args) {
	System.exit(TestHelpers.doTests(args, new Test8()));
    }
    

}

